sympy/core/numbers.py

```diff
@@ -2427,7 +2427,7 @@ def __new__(cls, expr, coeffs=None, alias=None, **args):
 
         if rep0 is not None:
             from sympy.polys.densetools import dup_compose
-            c = dup_compose(rep.rep, rep0.rep, dom)
+            c = dup_compose(rep.to_list(), rep0.to_list(), dom)
```

Comment:
Various places in the codebase access to the DMP.rep attribute. In this PR a DMP might be a DMP_Python which has the `.rep` attribute as its main internal data structure or it might be a DUP_Flint which has a completely different internal structure and does not have a .rep attribute.

The `.to_list()` method will work for either DMP_Python, DUP_Flint or DMP in older versions of SymPy. I could add a property method to DUP_Python like
```python
    @property
    def rep(self):
        return self.to_list()
```
That would be a bit more backwards compatible if there is any downstream code that manipulates DMP directly.

For now I haven't added that `.rep` property and instead I have just changed all places in the codebase that access `.rep` directly so that they use `.to_list()` instead. It is not clear to me that there is much need to maintain compatibility at this level. The DMP class is very clearly internal although it is available like `Poly(x).rep.rep` which does not involve any leading underscores.

Reply:
Maybe we should deprecate .rep in favor of ._rep

Reply:
I could add back the property but in such a way that it gives a deprecation warning and suggests to use `to_list()` instead. Using `.to_list()` already works and just returns `.rep` on master:
https://github.com/sympy/sympy/blob/c1b5f211c8442a390997bed6fe963ff9bb32875b/sympy/polys/polyclasses.py#L1527-L1529
It is potentially problematic that someone could modify this mutable list so it could perhaps be `return list(f.rep)` but as it stands in this PR I haven't changed the behaviour of this in `DMP_Python`.

Reply:
We might just need to think a bit more about how these APIs are used. I think at some point we will also want to add a SMP class for sparse polynomials. In that case, the .rep would be a completely different class. to_list could be implemented there, but unlike for DMP it would be an expensive operation. So it should be avoided unless you really do need the polynomial in list form. I expect any code that's using `.rep` is implicitly assuming the dense format when it shouldn't be. 

It's really unfortunate sparse polynomials weren't supported from the get go or else these issues wouldn't really exist.

Reply:
I don't think we could change it so that `Poly(x).rep.rep` gives anything other than a list without breaking something so I will ensure that it gives a list still but with a deprecation warning saying to use `.to_list()`.